### PR TITLE
* refactor(sdk): added option for custom metric collector for tune in…

### LINKED ISSUE #2402: [SDK] Support More MCKind in `tune`
### What you would like to be added?

Support specifying more types of metrics collector in the `tune` function.

https://github.com/kubeflow/katib/blob/51b246fa1c23db130c373fa9749897a0d0b334ab/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L189

### Why is this needed?

Currently, we only support `kind` param in `metrics_collector_config`, which means only `StdOut` and `Push` collector can be specified in `tune` function.

https://github.com/kubeflow/katib/blob/51b246fa1c23db130c373fa9749897a0d0b334ab/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L387-L391

However, there are many other collector types, such as `File`, `Prometheus`, `Custom`, `TensorFlowEvent` collector. They are important too. The `tune` function would be more flexible and powerful if users can specify all these kind of collectors.

Currently we support:

- [x] Push Collector: https://github.com/kubeflow/katib/pull/2369
- [x] StdOut Collector: https://github.com/kubeflow/katib/pull/2369
- [ ] File Collector
- [ ] Prometheus Collector
- [ ] Custom Collector
- [ ] TensorFlowEvent Collector

### Love this feature?

Give it a 👍 We prioritize the features with most 👍

### PR DESCRIPTION
**Pull request by prakhar479**
… katlib_client.py

Signed-off-by: prakhar479 <153047595+prakhar479@users.noreply.github.com>

added `custom_collector` field to `metrics_collector_config` in tune to allow for users to specify custom metrics collector for example `prometheus` 

fixes #2402

### CODE REVIEW

**Comment by prakhar479** [comment]
@Electronic-Waste can you review and let me know if any changes are required in this. Thanks a lot!

**Comment by Electronic-Waste** [comment]
@prakhar479 Yes, of course. Thanks for your great effort to Katib!

I'll look into this PR in the next few days.

**Review by andreyvelich — COMMENTED** [review]
/ok-to-test

**Comment by prakhar479** [comment]
I have corrected some oversights from my side and need approval for testing. Thanks!

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py [hunk 1]
```diff
@@ -251,8 +251,8 @@ def tune(
             pip_index_url: The PyPI url from which to install Python packages.
             metrics_collector_config: Specify the config of metrics collector, 
                 for example, `metrics_collector_config = {"kind": "Push"}`.
-                Currently, we only support `StdOut` and `Push` metrics collector.
-
+                for using custom metric collectors use "custom_collector" key,
+                for example, `metrics_collector_config = {"custom_collector": "prometheus"}`.
```
Comment (Electronic-Waste) [inline sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:255]: I guess `prometheus` is not a custom collector, since we define it here:

https://github.com/kubeflow/katib/blob/8eb0e86385a5b54d67fd775ff930f4f038b5b370/pkg/apis/controller/common/v1beta1/common_types.go#L216-L220

Here are some relevant resources which may be helpful for you:
https://github.com/kubeflow/katib/blob/master/examples/v1beta1/metrics-collector/custom-metrics-collector.yaml

Comment (Electronic-Waste) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:255]: Btw, `custom_collector` accepts `V1Container` as its input, not a `str`:

https://github.com/kubeflow/katib/blob/8eb0e86385a5b54d67fd775ff930f4f038b5b370/sdk/python/v1beta1/kubeflow/katib/models/v1beta1_collector_spec.py#L21-L43

Can you add some comments to remind users of this usage?

**Review by Electronic-Waste — COMMENTED** [review]
Nice job @prakhar479 , I left some comments for you!

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py (hunk 1 above)
Comment (Electronic-Waste) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:255]: Also, can you add some e2e tests like this (an example for `StdOut` Collector)?

https://github.com/kubeflow/katib/blob/8eb0e86385a5b54d67fd775ff930f4f038b5b370/test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py#L16

**Comment by Electronic-Waste** [comment]
PTAL👀 @andreyvelich @tenzen-y @johnugeorge when you have time.

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py (hunk 1 above)
Comment (prakhar479) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:255]: Thanks @Electronic-Waste, I have made the appropriate changes in the function comments to reflect the correct usage

for e2e tests, should I modify the existing test to use `metrics_collector_config` or should I create a separate test script for it altogether

Comment (Electronic-Waste) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:255]: @prakhar479 I think modifying the existing `run-e2e-tune-api.py` test is better since it's hard to pass the collector config to the test script: https://github.com/kubeflow/katib/blob/master/test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.sh. WDYT👀 @andreyvelich @tenzen-y @johnugeorge 

Also a small tip for reference: you can build images you need here

https://github.com/kubeflow/katib/blob/8eb0e86385a5b54d67fd775ff930f4f038b5b370/test/e2e/v1beta1/scripts/gh-actions/build-load.sh#L166-L170

**Comment by prakhar479** [comment]
I have modified the comment on usage of the custom metric param as well e2e test for tune Api.
For e2e test I have currently modified `build-load.sh` as suggested by @Electronic-Waste to build an image of custom metric using `dummy-collector.py` script and `Dockerfile.dummy-collector` file for building image for dummy collector container. Finally, I have modified `run-e2e-tune-api.py` adding the custom collector image as a V1 Container passed as a param to tune Api.

I was a bit confused with placement for these scripts and have placed all of them in `gh-action` directory. Let me know for any modifications, changes and fixes I need to make further.

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py (hunk 1 above)
Comment (prakhar479) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py]: Thanks @Electronic-Waste for the helpful insights 😄

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py [hunk 2]
```diff
@@ -251,8 +254,8 @@ def tune(
             pip_index_url: The PyPI url from which to install Python packages.
             metrics_collector_config: Specify the config of metrics collector, 
                 for example, `metrics_collector_config = {"kind": "Push"}`.
-                Currently, we only support `StdOut` and `Push` metrics collector.
```
Comment (Electronic-Waste) [inline sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:254]: I think we may need to tell users about the supported types of MC. So can you re-add this line?

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py [hunk 3]
```diff
@@ -251,8 +254,8 @@ def tune(
...
-
+                for using custom metric collectors use "custom_collector" key and provide instance of custom V1Container as value,
+                for example, `metrics_collector_config = {"kind" : "Custom", "custom_collector": <Instance of V1Container>}`.
```
Comment (Electronic-Waste) [inline sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:258]: Maybe we can reorganize these comments and explain each field in `metrics_collector_config`? Like

```
kind: specify the kind of Metrics Collector (currently we support...)
custom_collector: ...
```

File: test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py [hunk 4]
```diff
@@ -36,7 +37,19 @@ def objective(parameters):
...
+    metric_collector = V1Container(
+        name="dummy-collector",
+        image="dummy-collector:latest",
+        command=["python", "/app/dummy-collector.py"],
+        args=["--metric-name=result"],
+        env=[
+            client.V1EnvVar(name="EXPERIMENT_NAME", value=exp_name),
+            client.V1EnvVar(name="EXPERIMENT_NAMESPACE", value=exp_namespace)
+        ]
+    )
```
Comment (Electronic-Waste) [inline test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py:50]: I guess we can create another function `run_e2e_experiment_create_by_tune_custom` to run e2e tests for custom collector, rather than delete original e2e test for `StdOut` collector. Then we can run these e2e tests together in this file:)

WDYT👀 @prakhar479 @andreyvelich @tenzen-y @johnugeorge

**Review by Electronic-Waste — COMMENTED** [review]
@prakhar479 Sorry for my late response. I'm busy with other affairs these two weeks.

I left a few comments for you. Thanks for your greate contribution!

File: test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py (hunk 4 above)
Comment (prakhar479) [reply test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py]: I have made corresponding changes to run both e2e tests (one with custom metrics collector and another with default metric collector [StdOut])

**Comment by prakhar479** [comment]
I have made the neccesary changes that should also solve failing tests. Let me know about any further changes/suggestions @Electronic-Waste @andreyvelich @tenzen-y @johnugeorge.

File: test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py [hunk 5]
```diff
@@ -35,7 +96,7 @@ def objective(parameters):
         "b": search.double(min=0.1, max=0.2)
     }
 
-    # [3] Create Katib Experiment with 4 Trials and 2 CPUs per Trial.
+    # [4] Create Katib Experiment with 4 Trials and 2 CPUs per Trial.
```
Comment (Electronic-Waste) [inline test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py:99]: Why its numer is 4? I guess 3 is more suitable.

File: test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py [hunk 6]
```diff
@@ -78,11 +139,19 @@ def objective(parameters):
...
-        run_e2e_experiment_create_by_tune(katib_client, exp_name, exp_namespace)
+        exp_name = "tune-example-default-metrics-collector"
+        logging.info(f"Runnning E2E for Experiment created by tune: {exp_namespace}/{exp_name}")
+        run_e2e_experiment_create_by_tune_default_metrics_collector(katib_client, exp_name, exp_namespace)
+        logging.info("---------------------------------------------------------------")
+        logging.info(f"E2E is succeeded for Experiment created by tune: {exp_namespace}/{exp_name}")
+
+        exp_name = "tune-example-custom-metrics-collector"
+        logging.info(f"Runnning E2E for Experiment created by tune: {exp_namespace}/{exp_name}")
+        run_e2e_experiment_create_by_tune_custom_metrics_collector(katib_client, exp_name, exp_namespace)
```
Comment (Electronic-Waste) [inline test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py:154]: Maybe splitting these two e2e tests into separate `try-catch` clauses is better to identify error :)

File: test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py [hunk 7]
```diff
@@ -93,4 +162,4 @@ def objective(parameters):
         # Delete the Experiment.
         logging.info("---------------------------------------------------------------")
         logging.info("---------------------------------------------------------------")
-        katib_client.delete_experiment(exp_name, exp_namespace)
+        katib_client.delete_experiment(exp_name, exp_namespace)
```
Comment (Electronic-Waste) [inline test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py:165]: I think we need to delete the former experiment before we run another experiment. Otherwise, we may run into `xxx experiment alreay exists` error.

**Review by Electronic-Waste — COMMENTED** [review]
Thanks for your contribution @prakhar479 . I left some comments for you.

**Comment by prakhar479** [comment]
@Electronic-Waste I have fixed these issues lmk if anything else is needed. thanks!

**Comment by Electronic-Waste** [comment]
@prakhar479 Can you please fix the lint error and the error in tune API?

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py [hunk 8]
```diff
@@ -253,9 +256,18 @@ def tune(
...
                 executing the objective function.
             pip_index_url: The PyPI url from which to install Python packages.
-            metrics_collector_config: Specify the config of metrics collector,
-                for example, `metrics_collector_config = {"kind": "Push"}`.
-                Currently, we only support `StdOut` and `Push` metrics collector.
+            
+            `metrics_collector_config`: Specify the configuration for the metrics collector with following keys:
+            - **kind**: Specify the kind of Metrics Collector. Currently supported values are:
+                - `StdOut`: Collects metrics from standard output.
+                - `None`: No metrics collection.
```
Comment (andreyvelich) [inline sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:263]: This is not supported.

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py [hunk 9]
```diff
@@ -253,9 +256,18 @@ def tune(
...
+                - `File`: Writes metrics to a file.
+                - `TensorFlowEvent`: Collects metrics in TensorFlow Event format.
+                - `PrometheusMetric`: Exposes metrics in a Prometheus-compatible format.
+                - `Custom`: For custom metrics collection. Use the "custom_collector" key to specify the collector instance.
+                
+            - **custom_collector**: If the `kind` is set to `Custom`, you must provide an instance of a custom `V1Container` as the value. For example:
+                `metrics_collector_config = {"kind" : "Custom", "custom_collector": <Instance of V1Container>}`.
```
Comment (andreyvelich) [inline sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:270]: Can we add support for custom_collectors in the followup PRs when we get user requests ?
I feel that Data Scientists who will use `tune` API doesn't need such functionality.

File: test/e2e/v1beta1/scripts/gh-actions/Dockerfile.dummy-collector [hunk 10]
```diff
@@ -0,0 +1,9 @@
+FROM python:3.8-slim
+
+WORKDIR /app
+
+COPY test/e2e/v1beta1/scripts/gh-actions/dummy-collector.py .
+
+RUN pip install kubernetes
+
+CMD ["python", "dummy-collector.py"]
```
Comment (andreyvelich) [inline test/e2e/v1beta1/scripts/gh-actions/Dockerfile.dummy-collector:9]: Why do we need this container ?

File: test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py [hunk 11]
```diff
@@ -11,8 +12,68 @@
 # The default logging config.
 logging.basicConfig(level=logging.INFO)
 
+def run_e2e_experiment_create_by_tune_custom_metrics_collector(
```
Comment (andreyvelich) [inline test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py:15]: I don't think we need another E2E just to test this functionality.
We already run E2Es using these YAML files for various metrics collectors: https://github.com/kubeflow/katib/tree/master/examples/v1beta1/metrics-collector
I think for your feature, you just need to add unit tests for Katib Client to verify that Experiment has correct specification: https://github.com/kubeflow/katib/blob/master/sdk/python/v1beta1/kubeflow/katib/api/katib_client_test.py

**Review by andreyvelich — COMMENTED** [review]
Thank you for doing this @prakhar479!
I left a few comments.
cc @kubeflow/wg-automl-leads

File: test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py (hunk 11 above)
Comment (prakhar479) [reply test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py:15]: Sure! should I then remove the e2e test with custom metric collector from `run-e2e-tune-api.py`?

File: test/e2e/v1beta1/scripts/gh-actions/Dockerfile.dummy-collector (hunk 10 above)
Comment (prakhar479) [reply test/e2e/v1beta1/scripts/gh-actions/Dockerfile.dummy-collector:9]: This container is meant for serving as a dummy metric collector container intended for e2e test.

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py (hunk 9 above)
Comment (prakhar479) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py]: Sorry but can you elaborate on this a bit I am not quite sure what this intends since custom metric is already supported by underlying framework for tune function.

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py (hunk 8 above)
Comment (prakhar479) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:264]: Thanks for pointing this out. Can you lead me to where I can find all supported metric collector. For the current comment I had referenced `/pkg/ui/v1beta1/frontend/src/app/models/experiment.k8s.model.ts`

Comment (andreyvelich) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:264]: Yeah, UI also needs to be updated with the latest changes cc @Electronic-Waste 
Please ref the official CRDs APIs for Metrics Collector spec: https://github.com/kubeflow/katib/blob/master/pkg/apis/controller/common/v1beta1/common_types.go#L207-L227

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py (hunk 9 above)
Comment (andreyvelich) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py]: I meant that until we find use-cases when `tune` function needs to be used with `Custom` metrics collector, we can introduce it.
For example, I can see the value for File metrics collector, since users can write metrics into specific File during their training script, similar as for TensorFlow events.
However, for custom metrics collector, I can't see a use-cases when it might be useful with `tune` function.

Any thoughts @shannonbradshaw @prakhar479 @johnugeorge @tenzen-y ?

File: test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py (hunk 11 above)
Comment (andreyvelich) [reply test/e2e/v1beta1/scripts/gh-actions/run-e2e-tune-api.py:15]: Yes, please can you add the unit tests for your functionality: https://github.com/kubeflow/katib/blob/master/sdk/python/v1beta1/kubeflow/katib/api/katib_client_test.py ?

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py (hunk 8 above)
Comment (Electronic-Waste) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:264]: Thanks Andrey, I'll update UI with the latest changes.
